### DISCUSSION TITLE
Prism in Unity 3d

### DISCUSSION THREAD

**Discussion by nick5454** [PrismLibrary/Prism#2370] [General] [answered]
It would absolutely great for a micro version of Prism in Unity3d. Probably won't happen, but it would be nice.

**Comment by brianlagunas** [comment]
What would that even look like? What would it do, and what problems would it solve?

**Reply by nick5454** [reply]
In Unity3d there is a need for DI instead of always instantiating your own.

In my games I have to create singletons a lot ( GameController.Instance.Object can be a sprite, a method, a property )

There is also a concept of Packages ( basically a component with objects, image assets, code files ).

Typically when creating new games I package this up, but I still have to keep code that references GameController or other Controllers. If I had DI then I could just say GameObject(IGameController gameController) and I wouldn't have to worry about the bindings in the package.

Other 3rd party packages can also be imported and it would be nice to just register their specific scripts ( c# ) for DI

Navigation uses a Scene object ( Scenes.LoadScene )  and I could see how the NavigationService could override this.

Some game objects use MonoBehavior and a 3rd party Ioc allows you to use the Construct as an injection constructor ( Construct(IGameControler gameController) ) and injection happens there

When loading sprites ( Images, AnimationControllers, other components ) we have to do Resources.Load<Sprite>("path") and other redundant code like that located in the Resources folder. It would be nice to have an optimized Resolver to find these for us. GameObject.Find() is very expensive and I could see a need for drawing the objects to allow fast access.

There is a concept of Prefabs. This is a GameObject that has it's own hierarchy and is saved. Like a house with furniture, you edit it place all the objects, scripts and other items in it much like a component. But in controllers we often have to drag the Prefabs onto an Array to be included. This makes it difficult for reuse. If I have a package of, lets say a Chest that has random items ( Gold, Cake, Swords, etc ). I can't just package it because, the sprite that each represents has to be added in every game.

I also have an Object Pool that I created for use. Creating a Prefab at runtime is expensive so I have an ObjectPool that loads 100 monsters, 1000 particle effects, Walls, etc at the beginning and I then say GameController.Instance.GetObject(Objects.Goblin) to get the Monster set him to active and position him using a Transform. 

I'm making a lot of Hypercasual games and this would greatly increase the speed creating them. Another one exists but they take over the game, do checks, and just adds more complexity. It would be nice to just Add the Prism package, have a setup container, and import all of my package components and with them using DI for the service, objects, and ObjectPool manager and it would be very quick to create lots of games fast.

The object of a hypercasual game is to create it in less than 3 days.

Here's a link to one: https://assetstore.unity.com/packages/tools/utilities/extenject-dependency-injection-ioc-157735

But it seems to bulky, I would like a lightweight one and since I use Prism in all my Xamarin apps, it would be great to be in the Prism universe when making the games.

I'm more than happy to contribute to getting this started, but I have never worked on the inner workings of Prism.

Thanks,
Nick

**Reply by brianlagunas** [reply]
Most of what you describe is the need for DI. However, Prism is not DI, it just utilizes DI.  The core concepts in Prism do not work in Unity3d. Things such as Modularity, MVVM, Commanding, Event Aggregation (maybe), Navigation, etc. I have also been involved in a number of projects using Unity3d and I do not believe Prism is a fit for Unity3d development.

I would encourage you to start your own project and work towards your goal of creating a library that solves the problems you are running into.

**Comment by dansiegel** [comment, marked as answer]
So aside from finding some humor in packages called Prism.Unity.Unity3d and Prism.DryIoc.Unity3d.. ok maybe humor in how confused so many people would be by those names... you were right with your initial probably not going to happen. Now I will add here in full transparency that the "probably not going to happen" was also said of Prism for UWP. But let's look at why...

## Why Open Source Projects Exist

1. The Maintainers use it professionally, and thus it's existence helps to pay their bills and feed their family.
2. See # 1
3. The Maintainer was bored one weekend, started an experiment and wanted to see if they could use it and what people thought of it. Ultimately the goal is transition this to # 1 or abandon it.
4. Someone paid the maintainers good money to do it.
5. It helped them professionally... see # 1

## Why Open Source Get's Abandoned

Hey this is pretty relevant as many in our community know we did drop Prism for UWP when we released Prism 7. But a lot of folks don't understand why that might be. As noted in "Why Open Source Projects Exist", one of the former Prism core team members had a professional need for UWP. It was because of that, that you saw Prism for UWP. However as life changed for him, he was unable to maintain it. None of the other team members had a professional need for it. As a result it simply became a burden to get rid of as quick as possible.

Now to address an elephant in the room as the Prism team hasn't changed since that decision. Why Prism for Uno Platform, and why that doesn't suffer the same harsh fate. Prism for Uno Platform was accepted for a few primary reasons. First the implementation has very little to no actual code. Nearly all of Prism.Uno is referencing code from Prism.Wpf or Prism.Forms with compiler directives to change the namespace as needed. This means that while simply maintaining Prism.Wpf we automatically benefit Prism.Uno. Also we have gotten a commitment of support from the Uno team which themselves operate in the Open Source world and understand the burden of maintaining a large project.

## So Why Not Prism for Unity 3d???

Simply put neither @brianlagunas or myself have any financial interest in it. While Brian might like to play games to unwind and decompress, it may be possible he'd goof off with building a game on a live stream, but he isn't likely to do this professionally. As for me I'm just not a gamer. 

So where does that leave the idea? Well short of one of the following:
- someone paying a lot of money to make it happen
- someone from the community stepping up and becoming more than a fan of Prism and actually starting to take on some ownership of the project to a point that we ask them to join the team and trust them enough to let them add Prism for Unity 3d... 

It's not likely to ever happen.
